components/public-api-server/pkg/apiv1/workspace.go

```diff
@@ -77,6 +77,52 @@ func (s *WorkspaceService) GetWorkspace(ctx context.Context, req *connect.Reques
 	}), nil
 }
 
+func (s *WorkspaceService) StreamWorkspaceStatus(ctx context.Context, req *connect.Request[v1.StreamWorkspaceStatusRequest], stream *connect.ServerStream[v1.StreamWorkspaceStatusResponse]) error {
+	workspaceID, err := validateWorkspaceID(req.Msg.GetWorkspaceId())
+	if err != nil {
+		return err
+	}
+
+	logger := ctxlogrus.Extract(ctx).WithField("workspace_id", workspaceID)
+
+	conn, err := getConnection(ctx, s.connectionPool)
+	if err != nil {
+		return err
+	}
+
+	workspace, err := conn.GetWorkspace(ctx, workspaceID)
+	if err != nil {
+		logger.WithError(err).Error("Failed to get workspace.")
+		return proxy.ConvertError(err)
+	}
+
+	if workspace.LatestInstance == nil {
+		logger.WithError(err).Error("Failed to get latest instance.")
+		return connect.NewError(connect.CodeFailedPrecondition, fmt.Errorf("instance not found"))
+	}
+
+	ch, err := conn.InstanceUpdates(ctx, workspace.LatestInstance.ID)
+	if err != nil {
+		logger.WithError(err).Error("Failed to get workspace instance updates.")
+		return proxy.ConvertError(err)
+	}
+
+	for update := range ch {
+		instance, err := convertWorkspaceInstance(update, workspace.Workspace.Shareable)
+		if err != nil {
+			logger.WithError(err).Error("Failed to convert workspace instance.")
+			return proxy.ConvertError(err)
+		}
+		_ = stream.Send(&v1.StreamWorkspaceStatusResponse{
```

Comment:
Handle error from send. When that happens, you'll want to terminate the request - return a server-side error.

Unsure what will happen with the underlying workspace instance stream, but I'm hoping the context will get cancelled and it will get cleaned up.